### PR TITLE
[AIR] Change `FeatureHasher` input schema

### LINKED ISSUE #27505: [AIR] `FeatureHasher` doesn't work the way it's supposed to
### What happened + What you expected to happen

I expected `FeatureHasher` to work like [sklearn's `FeatureHasher`](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.FeatureHasher.html). 

**What it does right now**
`FeatureHasher` expects columns to contain individual tokens.
```
    token1       token2         token3   
0      I           like         Python
1      I         dislike        Python
```

**What it should do**
`FeatureHasher` should expect a table that describes token counts

```
   I  like  dislike  Python
0  1     1        0       1
1  1     0        1       1
```

**Why we should make this change**
* Consistent with sklearn
* Allows you to chain `CountVectorizer` with `FeatureHasher`
* Current implementation isn't useful; `HashingVectorizer` does basically the same thing.

### Versions / Dependencies

**Ray**: 11abc8974659bfb70dd393f0d21f46cce1ed16cf

### Reproduction script

```
import pandas as pd
import ray
from ray.data.preprocessors import FeatureHasher


# This dataframe represents the counts from the documents "I like Python" and "I
# dislike Python".
df = pd.DataFrame({
    "I": [1, 1],
    "like": [1, 0],
    "dislike": [0, 1],
    "Python": [1, 1]
})
ds = ray.data.from_pandas(df)

hasher = FeatureHasher(["I", "like", "dislike", "Python"], num_features=8)
print(hasher.fit_transform(ds).to_pandas().to_numpy())
```

This document-term matrix isn't useful.
```
[[0 0 0 3 1 0 0 0]
 [0 0 0 3 1 0 0 0]]
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

See #27505.

## Related issue number

Fixes #27505 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/preprocessors/hasher.py [hunk 1]
```diff
@@ -33,17 +80,12 @@ def __init__(self, columns: List[str], num_features: int):
...
             for column in self.columns:
-                hashed_value = simple_hash(row[column], self.num_features)
-                hash_counts[hashed_value] = hash_counts[hashed_value] + 1
-            return {
-                f"hash_{joined_columns}_{i}": hash_counts[i]
-                for i in range(self.num_features)
-            }
+                hashed_value = simple_hash(column, self.num_features)
+                hash_counts[hashed_value] += row[column]
+            return {f"hash{i}": hash_counts[i] for i in range(self.num_features)}
```
Comment (matthewdeng) [inline python/ray/data/preprocessors/hasher.py:88]:
```suggestion
            return {f"hash_{i}": hash_counts[i] for i in range(self.num_features)}
```

**Review by matthewdeng — APPROVED** [review]

File: python/ray/data/preprocessors/hasher.py (hunk 1 above)
Comment (bveeramani) [reply python/ray/data/preprocessors/hasher.py]: Fixed

File: python/ray/data/preprocessors/hasher.py [hunk 2]
```diff
@@ -9,19 +9,68 @@
...
+        ``num_features``.
+
+    .. seealso::
+        :class:`~ray.data.preprocessors.CountVectorizer`
+            Use this preprocessor to generate inputs for :class:`FeatureHasher`.
+
+        :class:`ray.data.preprocessors.HashingVectorizer`
+            If your input data describes documents rather than token frequencies,
+            use :class:`~ray.data.preprocessors.HashingVectorizer`.
+    """  # noqa: E501
```
Comment (richardliaw) [inline python/ray/data/preprocessors/hasher.py:73]: nice ! we should do this way more.

**Review by richardliaw — APPROVED** [review]

**Review by clarkzinzow — APPROVED** [review]
LGTM!

**Pull request merged**